File: java/src/main/java/noie/linmimeng/noiedigitalsystem/api/adapter/V3EventV1Adapter.java
package noie.linmimeng.noiedigitalsystem.api.adapter;

import java.util.Objects;
import noie.linmimeng.noiedigitalsystem.api.proto.event.v1.Cursor;

/**
 * v3 Event (v1 package) Adapter - minimal helpers for streaming primitives.
 *
 * @since 3.0.0
 */
public final class V3EventV1Adapter {

    private V3EventV1Adapter() {
        // [Index] NDS-JAVA-V3-EVENTV1-001 [Constraint] Utility class; instantiation is prohibited.
    }

    /**
     * Create Cursor from opaque bytes.
     *
     * @param value opaque bytes (must be non-empty)
     * @return Cursor
     * @throws IllegalArgumentException if value is null/empty
     * @since 3.0.0
     */
    public static Cursor createCursor(byte[] value) {
        if (value == null || value.length == 0) {
            throw new IllegalArgumentException("value must be non-empty");
        }
        return Cursor.newBuilder()
            .setValue(com.google.protobuf.ByteString.copyFrom(value))
            .build();
    }

    /**
     * Extract opaque bytes from Cursor.
     *
     * @param cursor Cursor (non-null)
     * @return bytes
     * @since 3.0.0
     */
    public static byte[] toBytes(Cursor cursor) {
        Objects.requireNonNull(cursor, "cursor");
        return cursor.getValue().toByteArray();
    }
}

File: csharp/src/Noie.Nds.Api/Adapter/V3EventV1Adapter.cs
using Nds.Event.V1;

namespace Noie.Nds.Api.Adapter;

/// <summary>
/// [Index: NDS-CSHARP-V3-EVENTV1ADAPTER-000] v3 event (v1 package) adapter.
///
/// <para>
/// [Semantic] Minimal helpers for v3+ streaming primitives (Cursor).
/// </para>
/// </summary>
public static class V3EventV1Adapter
{
    /// <summary>
    /// [Index: NDS-CSHARP-V3-EVENTV1ADAPTER-010] Create a <see cref="Cursor"/> from opaque bytes.
    /// </summary>
    /// <exception cref="ArgumentException">If <paramref name="value"/> is null or empty.</exception>
    public static Cursor CreateCursor(byte[] value)
    {
        if (value is not { Length: > 0 })
            throw new ArgumentException("value must be non-empty", nameof(value));

        return new Cursor { Value = Google.Protobuf.ByteString.CopyFrom(value) };
    }

    /// <summary>
    /// [Index: NDS-CSHARP-V3-EVENTV1ADAPTER-011] Extract opaque bytes from a <see cref="Cursor"/>.
    /// </summary>
    public static byte[] ToBytes(Cursor cursor)
    {
        ArgumentNullException.ThrowIfNull(cursor);
        return cursor.Value.ToByteArray();
    }
}

